src/agent/results.rs

```diff
@@ -0,0 +1,70 @@
+use agent::api::AgentApi;
+use crates::{Crate, GitHubRepo};
+use errors::*;
+use ex::Experiment;
+use log;
+use results::{TestResult, WriteResults};
+use std::io::Read;
+use std::ops::DerefMut;
+use std::sync::{Arc, Mutex};
+use toolchain::Toolchain;
+
+#[derive(Clone)]
+pub struct ResultsUploader<'a> {
+    api: &'a AgentApi,
+    shas: Arc<Mutex<Vec<(GitHubRepo, String)>>>,
+}
+
+impl<'a> ResultsUploader<'a> {
+    pub fn new(api: &'a AgentApi) -> Self {
+        ResultsUploader {
+            api,
+            shas: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+}
+
+impl<'a> WriteResults for ResultsUploader<'a> {
+    fn already_executed(
+        &self,
+        _ex: &Experiment,
+        _toolchain: &Toolchain,
+        _krate: &Crate,
+    ) -> Result<Option<TestResult>> {
+        // TODO: not yet implemented
+        Ok(None)
+    }
+
+    fn record_sha(&self, _ex: &Experiment, repo: &GitHubRepo, sha: &str) -> Result<()> {
+        self.shas
+            .lock()
+            .unwrap()
+            .push((repo.clone(), sha.to_string()));
+        Ok(())
+    }
+
+    fn record_result<F>(
+        &self,
+        _ex: &Experiment,
+        toolchain: &Toolchain,
+        krate: &Crate,
+        f: F,
+    ) -> Result<TestResult>
+    where
+        F: FnOnce() -> Result<TestResult>,
+    {
+        let mut log_file = ::tempfile::NamedTempFile::new()?;
+        let result = log::redirect(log_file.path(), f)?;
```

Comment:
I'm not experienced with slog, is there a way to capture the logs without storing them on disk? I remember a pointless fight with lifetimes when I tried it.

Reply:
I've not used slog before, but I can see why you're having issues - to be used in a logger, `Drain`s in general need to be `Send + Sync` (since they could be used by multiple threads). So let's say you fixed this by add a `Mutex`. However, AFAIK there's no safe way to actually move something between threads that's `Send` but not `'static` (which makes sense), so you need a `'static` bound as well.

In the specific case of `slog_scope` you possibly could relax the bound, but you need to be sure that no other threads have still got a reference to the value when you exit the scope - sounds a bit tricky to me.

So this seems fine. And in time if we want to stream from a `Reader` rather than sending the `Vec` directly (due to memory constraints), this works well as well.